support external queue implementations

psychocoderHPC opened the pull request:
solves #820

By providing a queue interface class it is possible to handle user defines queues within alpaka.

- add queue interface `ICpuQueue`
- remove usage of hardcoded register function for each queue type within the device

The second commit provides a OpenMP collective queue as test case for an externally defined queue which is designed with the requirements of #820 in mind.

psychocoderHPC committed:
support for queues defined outside of alpaka

By providing a queue interface class it is possible to handle user defines queues within alpaka.

- add queue interface `ICpuQueue`
- remove usage of hard coded register function for each queue type within the device

psychocoderHPC commented:
CC-ing: @krzikalla

psychocoderHPC commented:
As I wrote in #820 I do not think that the highly specific collective queue implementation fitts into the generic concepts for alpaka. So IMO this is the way how to support the user requirements.

BenjaminW3 commented on test/unit/queue/src/QueueTest.cpp:
```diff
@@ -159,7 +161,27 @@ void operator()()
+        std::tuple<alpaka::dev::DevHipRt, alpaka::queue::QueueHipRtBlocking>,
+        std::tuple<alpaka::dev::DevHipRt, alpaka::queue::QueueHipRtNonBlocking>
+#endif
+    >;
```
Why did you remove it completely from `test/queue/Queue.hpp`? I would prefer to still have the `TestQueues` there and only appending the `QueueCpuOmp2Collective` here.

BenjaminW3 replied:
Or simply add it to `TestQueues` there.

psychocoderHPC replied:
I moved it to the tests because in this file defines traits like `IsBlockingQueue`. This make it impossible to have all trait definitions outside of the file `Queue.hpp` but I do not like to define the traits within `Queue.hpp` because I like to have clear borders between the "user code` (the collective queue) and alpaka.

BenjaminW3 replied:
But then please leave the alpaka-queues in there and append the new one.

BenjaminW3 commented on test/unit/event/src/EventTest.cpp:
```diff
@@ -285,7 +286,28 @@ void operator()()
+        std::tuple<alpaka::dev::DevHipRt, alpaka::queue::QueueHipRtBlocking>,
+        std::tuple<alpaka::dev::DevHipRt, alpaka::queue::QueueHipRtNonBlocking>
+#endif
+    >;
```
Same as in `QueueTest.cpp`

BenjaminW3 commented on include/alpaka/queue/cpu/ICpuQueue.hpp:
```diff
@@ -0,0 +1,46 @@
+                ICpuQueue() = default;
+                //! enqueue the event into the given queue ------------------------------------
+                virtual void enqueue(std::shared_ptr<ICpuQueue> &, event::EventCpu &) = 0;
+                //! the given queue is waiting for the event-----------------------------------
```
Could you please use the current style with a separate line for the comment. This would make it easier to mass-replace them later.

psychocoderHPC replied:
^^ I do not find examples for the style but think I got what you mean

BenjaminW3 commented on include/alpaka/queue/cpu/ICpuQueue.hpp:
```diff
@@ -0,0 +1,46 @@
+ */
+
+
+#pragma once
```
Could you please remove the second new-line?

BenjaminW3 commented on include/alpaka/queue/QueueCpuNonBlocking.hpp:
```diff
@@ -73,6 +74,20 @@ namespace alpaka
+                    //-----------------------------------------------------------------------------
+                    void enqueue(std::shared_ptr<cpu::ICpuQueue> & iQueue, event::EventCpu & ev) final
+                    {
+                        auto spQueueImpl = std::static_pointer_cast<QueueCpuNonBlockingImpl>(iQueue);
```
Do we really need the separate `iQueue` parameter? We should have a `enqueue` and `wait` overload for the non-shared_ptr types. Then we only would have to pass `*this`.

psychocoderHPC replied:
I will try it again. I started with that before but somehow I was unhappy with that. I do not know anymore why so I will try it again.

BenjaminW3 commented:
There is another option to the interface inheritance. We could probably give the `DevCpu` a template parameter for additional queues it has to support. It then would hold a vector of variants of the built-in queues and the new queues. However, I do not think that this would be better.

BenjaminW3 commented on include/alpaka/queue/QueueCpuBlocking.hpp:
```diff
@@ -217,3 +232,5 @@ namespace alpaka
     }
 }
+
+#include <alpaka/event/EventCpu.hpp>
```
Why is this strange include necessary?

psychocoderHPC replied:
This is needed to pass the header checks where we check that a header is including all dependencies e.g. https://github.com/ComputationalRadiationPhysics/alpaka/pull/822/files#diff-5b4348d85ae1c4f4c1c432f2aa3b97bdR74 will call wait with an event but the specialization for the meaning of wait is within `EventCpu.hpp`. But since event depends on queue and here we depend of knowing the event we need to pull in the specializations.

BenjaminW3 commented on test/unit/queue/src/CollectiveQueue.cpp:
```diff
@@ -0,0 +1,174 @@
+
+TEST_CASE( "queueCollective", "[queue]")
+{
+    TestCollective{}();
```
Do we need the separate class at all? It is only used in the templatized tests as workaround. We should be able to move the code directly in here.

BenjaminW3 commented on test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp:
```diff
@@ -0,0 +1,454 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+
```
Please remove the second newline here as well.

BenjaminW3 commented on test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp:
```diff
@@ -0,0 +1,454 @@
+#pragma once
+
+#ifdef ALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLED
+
```
Do we need the check here similar to all other OMP dependant files?
```
#if _OPENMP < 200203
    #error If ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLED is set, the compiler has to support OpenMP 2.0 or higher!
#endif
```

BenjaminW3 commented on test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp:
```diff
@@ -0,0 +1,454 @@
+                //#############################################################################
+                //! The blocking queue trait specialization for a OpenMP2 collective CPU queue.
+                template<>
+                struct IsBlockingQueue<
```
Could we move the `IsBlockingQueue` trait (maybe as `IsBlocking`) for this and all the other queues to the main `alpaka` library?
Could we please rename this queue to `QueueCpuBlockingOmp2Collective` so hat it matches the style of all others?

psychocoderHPC replied:
> Could we please rename this queue to `QueueCpuBlockingOmp2Collective` so that it matches the style of all others?

I also thought at this name before but it is not blocking within a parallel region so I avoided to use blocking within the name

psychocoderHPC replied:
> Could we move the `IsBlockingQueue` trait (maybe as `IsBlocking`) for this and all the other queues to the main `alpaka` library?

We can do it but outside of this PR.

BenjaminW3 replied:
But why does `IsBlockingQueue` then return true? A queue that is blocking is defined by the fact that a call to `queue::enqueue(queue, task);` (which is also hidden within `kernel::exec`) does not return before the task is executed. Is this true or false for this queue?

psychocoderHPC replied:
because it depends where you use it (in a parallel region or not) outside it is blocking and to not move over tests those require a nonblocking queue it is set to blocking.

psychocoderHPC committed:
add test case for a user defined queue

- To allow the specialization of traits needed for test cases this Commit move the list of all used queue types for tests to the coresponding test.
- provide a user defined queue `iQueueCpuOmp2Collective`
- add queue tests for `QueueCpuOmp2Collective`

psychocoderHPC committed:
remove queue shared pointer specializations

remove trait shared pointer specilization for queues.

BenjaminW3 commented on include/alpaka/queue/cpu/ICpuQueue.hpp:
```diff
@@ -0,0 +1,45 @@
+            {
+            public:
+                //-----------------------------------------------------------------------------
+                ICpuQueue() = default;
```
Do we really need the explicitly defaulted default construcor? I think this is not necessary for an abstract interface class.

BenjaminW3 commented on include/alpaka/queue/QueueCpuNonBlocking.hpp:
```diff
@@ -73,6 +74,18 @@ namespace alpaka
                     //-----------------------------------------------------------------------------
                     ~QueueCpuNonBlockingImpl() = default;
```
We should now remove this destructor as it is missing the virtual keyword and it now has a base class.

BenjaminW3 commented on include/alpaka/queue/QueueCpuBlocking.hpp:
```diff
@@ -59,6 +60,18 @@ namespace alpaka
                     //-----------------------------------------------------------------------------
                     ~QueueCpuBlockingImpl() = default;
```
We should now remove this destructor as it is missing the virtual keyword and it now has a base class.

BenjaminW3 commented on test/common/include/alpaka/test/queue/QueueCpuOmp2Collective.hpp:
```diff
@@ -0,0 +1,455 @@
+                    //-----------------------------------------------------------------------------
+                    auto operator=(QueueCpuOmp2CollectiveImpl &&) -> QueueCpuOmp2CollectiveImpl & = delete;
+                    //-----------------------------------------------------------------------------
+                    ~QueueCpuOmp2CollectiveImpl() = default;
```
Here as well

psychocoderHPC committed:
github annotations

- remove destructor in parent class
- avoid clang diagnistic warning for the pure interface class
- fix lambda usage

BenjaminW3 commented:
CI complains: `CollectiveQueue.cpp(57): error: __host__ or __device__ annotation on lambda requires --expt-extended-lambda nvcc flag`
